### DISCUSSION TITLE
[@types/alpinejs] Make `AlpineComponent` a `class` instead of `type`

### DISCUSSION THREAD

**Discussion by msrumon** [DefinitelyTyped/DefinitelyTyped#69516] [Issues with a @types package]
Consider the following scenario:

```typescript
class App {
  //
}

Alpine.data('app', new App())
```

Here, we can't utilize `AlpineComponent` by any means. If it was a class, we could use `extends` and be happy. Currently, we're getting `'AlpineComponent' only refers to a type, but is being used as a value here.` error if we use it like:

```typescript
import { AlpineComponent } from 'alpinejs'

class App extends AlpineComponent<any> {
  //
}

Alpine.data('app', new App())
```

[a comment by an automated account is not shown]

**Comment by ekwoka** [comment]
Thanks for bringing this up!


So multiple things.

1. The syntax shown wouldn't really work, since you need to pass a constructor function to data, not a class isntance.

2. Until [3.13.6](https://github.com/alpinejs/alpine/releases/tag/v3.13.6), Alpine didn't support classes in expression for accessing instance methods ([I made the PR that allowed it](https://github.com/alpinejs/alpine/pull/4038)). The last update to this part of the code was before that.

3. I am aware of issues using classes and it should be solved as of #67881  . Can you also check again with proper syntax and updated packages to ensure the error still happens? Can you provide the specific errors?

4. Specifically making `AlpineComponent` a `class` or `abstract class` is not possible here, as that is runtime code. This package here only involves Types for static type checking, not runtime semantics. You can use `class MyData implements AlpineComponent` to get some of the hints and help. Though I believe `this` will not have proper typings for magics and stuff. Tricky Tricky.

I have a few updates to the types I have planned, and hopefully this week (maybe even weekend) I can get a PR going for them.

**Reply by msrumon** [reply]
(3) => Error is already mentioned above while using `extends`. I'm using the latest version as of now.

(4) => Yes, `implements` works. But as you said, `this` has NO magic properties and methods. If you can make some adjustments in the type definitions so that those magics are available on `this` object, that'd be awesome. And I think the only way to do it is by making `AlpineComponent` a class.

Anyway, I'm not very knowledgeable in TS types. Maybe it's possible without making it a class. You know better.

> Last but not least, thanks for the alpinejs/alpine#4038.

**Reply by ekwoka** [reply]
AlpineComponent cannot be a class.

Classes are runtime. They exist at the value level.

This is type definitions. They only exist at the type level.

If we say the class exists, you won't get TS errors, but your bundler will complain since there is no actual class to use as a value.

I am aware of the `this` issues and getting classes working is something I want, but thats not a solution.

Not unless Alpine actually adds a class for it.

I'm hoping to find a way to get classes to work better with `ThisType` but it may be a more fundamental limitation. Or there might be some other kind of userland TS that might need to be added, not sure.
